src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/NewLineAnalyzer.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class NewLineAnalyzer: DiagnosticAnalyzer
+    {
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(
+                analysisContext =>
+                {
+                    analysisContext.RegisterSyntaxTreeAction(
+                        nodeContext => {
+                            var rootNode = nodeContext.Tree.GetCompilationUnitRoot(nodeContext.CancellationToken);
+                            foreach (var node in rootNode.DescendantTokens())
+                            {
+                                CheckNewLines(node.LeadingTrivia, nodeContext);
+                                CheckNewLines(node.TrailingTrivia, nodeContext);
+                            }
+                        });
+                });
+        }
+
+        private static void CheckNewLines(SyntaxTriviaList syntaxTriviaList, SyntaxTreeAnalysisContext nodeContext)
+        {
+            int newLineCount = 0;
+            SyntaxTrivia previousTrivia = default;
+            foreach (var trivia in syntaxTriviaList)
+            {
+                if (trivia.IsKind(SyntaxKind.EndOfLineTrivia))
+                {
+                    if (previousTrivia.IsKind(SyntaxKind.WhitespaceTrivia))
+                    {
+                        nodeContext.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0102, previousTrivia.GetLocation()));
+                    }
+
+                    if (newLineCount > 0)
+                    {
+                        nodeContext.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0101, trivia.GetLocation()));
+                    }
+
+                    newLineCount++;
+                }
+                else if (!trivia.IsKind(SyntaxKind.WhitespaceTrivia))
+                {
+                    // Reset new line count on comments and other stuff
+                    newLineCount = -1;
+                }
+
+                previousTrivia = trivia;
+            }
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(new[]
+        {
+            Descriptors.AZC0101,
+            Descriptors.AZC0102
+        });
+    }
+}
```

Comment:
Any concerns about there not being a newline at the EOF?

Reply:
We are not really enforcing that but maybe we should.